git-branchless/src/revset/grammar.lalrpop

```diff
@@ -1,110 +1,97 @@
-    <a:Expr3> <op:BinOp2> <b:Expr2> => Expr::Fn(op.into(), vec![a, b]),
-    Expr2,
-};
-
-BinOp2: &'static str = {
-    "&" => "intersection",
-    "and" => "intersection",
-    "-" => "difference",
-    "%" => "only",
-};
-
-Expr2: Expr = {
-    <a:Expr2> <op:BinOp1> <b:Expr15> => Expr::Fn(op.into(), vec![a, b]),
-    Expr15,
-};
-
-BinOp1: &'static str = {
-    ":" => "range",
-    ".." => "range",
-};
-
-Expr15: Expr = {
-    <pre:(<Prefix>)*> <e:Expr1> => {
-        let mut e = e;
-        for v in pre { e = Expr::Fn(v.into(), vec![e]); }
-        e
-    },
+// Below implements a hierarchy of operator precedences. The lower-numbered
+// `Expr`s bind less tightly than the higher-numbered `Expr`s.
+
+pub Expr: Expr<'input> = {
+    "(" <Expr> ")",
+    <lhs:Expr> "|"  <rhs:Expr2> => Expr::FunctionCall(Cow::Borrowed("union"), vec![lhs, rhs]),
+    <lhs:Expr> "+"  <rhs:Expr2> => Expr::FunctionCall(Cow::Borrowed("union"), vec![lhs, rhs]),
+    <lhs:Expr> "or" <rhs:Expr2> => Expr::FunctionCall(Cow::Borrowed("union"), vec![lhs, rhs]),
+    <Expr2>,
 }
 
-Expr1: Expr = {
-    <e:Expr0> <post:(<Postfix>)*> => {
-        let mut e = e;
-        for v in post { e = Expr::Fn(v.into(), vec![e]); }
-        e
-    },
+Expr2: Expr<'input> = {
+    <lhs:Expr2> "&"   <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("intersection"), vec![lhs, rhs]),
+    <lhs:Expr2> "and" <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("intersection"), vec![lhs, rhs]),
+    <lhs:Expr2> "-"   <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("difference"),   vec![lhs, rhs]),
+    <lhs:Expr2> "%"   <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("only"),         vec![lhs, rhs]),
+    <Expr3>
 }
 
-Expr0: Expr = {
-    Symbol2 => Expr::Name(<>),
-    <f:Symbol1> "(" <args:(Expr ",")*> <last:Expr?> ")" => {
-        // Function call.
-        let mut arg_list: Vec<Expr> = args.into_iter().map(|(e, _)| e).collect();
-        if let Some(last_arg) = last { arg_list.push(last_arg); }
-        Expr::Fn(f.into(), arg_list)
-    },
-    "(" <Expr> ")"
-};
+Expr3: Expr<'input> = {
+    <lhs:Expr3> ":"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("range"),       vec![lhs, rhs]),
+    <lhs:Expr3> ":"              =>  Expr::FunctionCall(Cow::Borrowed("descendants"), vec![lhs,    ]),
+                ":"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("ancestors"),   vec![     rhs]),
 
-Prefix: &'static str = {
-    // negation unimplemented for now
-    // "!" => "negate",
-    // "not " => "negate",
+    // For Mercurial users' familiarity.
+    <lhs:Expr3> "::"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("range"),       vec![lhs, rhs]),
+    <lhs:Expr3> "::"              =>  Expr::FunctionCall(Cow::Borrowed("descendants"), vec![lhs,    ]),
+                "::"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("ancestors"),   vec![     rhs]),
 
-    "::" => "ancestors",
+    <lhs:Expr3> ".." <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("only"), vec![lhs, rhs]),
```

Comment:
This is out of scope for this PR, but it's similar to my other comment and may be easy to add while you're here: would you consider adding a three dot notation, again for familiarity/parity with `git`s native `...` notation? If I'm not mistaken, it would be something like `union(only(lhs, rhs), only(rhs, lhs))`.

To be fair, though, I'm personally ambivalent to triple dot ... I don't think I've ever used it. It just strikes me as something that a `git` might user might expect to work out of the box when first learning `git branchless`.

Reply:
I'm open to it, but I've never actually used `...` either, so I'm going to defer it for now 😄.

Reply:
Makes sense to me; thanks for considering!